client.c

```diff
@@ -299,6 +299,22 @@ static int call0 (FTN_NODE *node, BINKD_CONFIG *config)
 #endif
   struct addrinfo *ai, *aiNodeHead, *aiHead, hints;
   int aiErr;
+  static int sa_init = 1;
+  static struct sockaddr_in sa_0;
+#ifdef AF_INET6
+  static struct sockaddr_in6 sa6_0;
+#endif
+
+  if (sa_init)
+  {
+    sa_init = 0;
+    memset((void *)&sa_0, 0, sizeof(struct sockaddr_in));
+    sa_0.sin_family = AF_INET;
```

Comment:
Hi Pavel,

On Mon, Jan 18, 2016 at 4:54 PM, Pavel <notifications@github.com> wrote:

> In client.c <https://github.com/pgul/binkd/pull/4#discussion_r50013576>:
>
> > @@ -299,6 +299,22 @@ static int call0 (FTN_NODE *node, BINKD_CONFIG *config)
> >  #endif
> >    struct addrinfo *ai, *aiNodeHead, *aiHead, hints;
> >    int aiErr;
> > +  static int sa_init = 1;
> > +  static struct sockaddr_in sa_0;
> > +#ifdef AF_INET6
> > +  static struct sockaddr_in6 sa6_0;
> > +#endif
> > +
> > +  if (sa_init)
> > +  {
> > +    sa_init = 0;
> > +    memset((void *)&sa_0, 0, sizeof(struct sockaddr_in));
> > +    sa_0.sin_family = AF_INET;
>
> This code is not thread-safe.
> We may catch race condition if both threads check sa_init, then both set
> sa_init to 0, then:
> 1st: memset(), sa_0.sin_family = AF_INET
> 2nd: memset()
> 1st: running further with zero sin_family.
>
​A, yes, of course, I didn't consider multitasking...

I think we have to init these structures before branch, may be in
> do_client() or clientmgr().
>
​I'll look into it...

-- 
Met vriendelijke groet / Best regards,
Wilfred van Velzen